app/commands/submission/test_run/process.rb

```diff
@@ -72,30 +57,73 @@ def handle_error!
       end
 
       def cancel_other_services!
+        return if solution_head_run?
+
         ToolingJob::Cancel.(submission.uuid, :analyzer)
         ToolingJob::Cancel.(submission.uuid, :representer)
       end
 
-      memoize
-      def submission
-        Submission.find_by!(uuid: tooling_job.submission_uuid)
+      def update_status!(status)
+        solution_head_run? ? update_solution_status!(status) : update_submission_status!(status)
       end
 
-      def update_status!(status)
+      def update_submission_status!(status)
         submission.with_lock do
           return if submission.tests_cancelled?
 
           submission.send("tests_#{status}!")
         end
       end
 
+      def update_solution_status!(status)
+        submission.solution.with_lock do
+          return unless submission.exercise.git_sha == git_sha
+
+          submission.solution.send("published_iteration_head_tests_status_#{status}!")
+        end
```

Comment:
Is the lock here necessary? It looks like the check inside (`submission.exercise.git_sha == git_sha`) is unrelated to the locked object (`submission.solution`).

Reply:
There's a race here where the exercise's git sha could change, but I guess if that happens another job will get scheduled, so it can probably be safely removed.